File: algorithms/binaryseach/binary_search.py
def binary_search_strict(numbers: list, value: int) -> int:
    """
    二分探索:一致する値のインデントを返却する。見つからない場合は-1を返却する。
    """
    def _binary_search(numbers: list, value: int,
                       left: int, right: int) -> int:
        if left > right:
            return -1

        middle = (left + right) // 2
        if numbers[middle] == value:
            return middle
        elif numbers[middle] < value:
            return _binary_search(numbers, value, middle + 1, right)
        else:
            return _binary_search(numbers, value, left, middle - 1)

    return _binary_search(numbers, value, 0, len(numbers) - 1)


def binary_search(numbers: list, value:int) -> int:
    """
    二分探索:一致する値のインデントを返却する。見つからない場合は、絶対値が一番小さい値のインデックスを返却する。
    """
    left, right = 0, len(numbers)-1
    while left <= right:
        mid = (left + right) // 2
        if numbers[mid] == value:
            return mid
        elif numbers[mid] < value:
            left = mid + 1
        else:
            right = mid - 1
    idx1 = min(left, len(numbers)-1)
    idx2 = max(right, 0)
    l = abs(numbers[idx1] - value)
    r = abs(numbers[idx2] - value)
    return idx1 if l < r else idx2

File: algorithms/java/search/BinarySearch.java
package algorithms.java.search;

public class BinarySearch {

  /**
   * Binary Search.
   * 
   * @param nums   a sorted array of distinct integers
   * @param target a target value
   * @return return the index if the target is found. If not, return the index
   *         where it would be if it were inserted in order
   */
  public int searchInsert(int[] nums, int target) {
    return searchInsert(nums, target, 0, nums.length - 1);
  }

  private int searchInsert(int[] nums, int target, int left, int right) {
    if (left > right) {
      // 見つからない場合、挿入すべきINDEXを返却
      int r = Math.max(0, right);
      return nums[r] < target ? r + 1 : r;
    }
    int mid = (left + right) / 2;
    if (nums[mid] == target) {
      return mid;
    }
    if (nums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
    return searchInsert(nums, target, left, right);
  }
}
